File: internal/crypto/nip44.go
package crypto

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"io"

	secp "github.com/decred/dcrd/dcrec/secp256k1/v4"
	"golang.org/x/crypto/chacha20poly1305"
	"golang.org/x/crypto/hkdf"
)

const (
	nip44Version    = 0x02
	nonceSize       = 32
	keyMaterialSize = 76
	keySize         = 32
	ivSize          = 12
)

func sharedSecretECDH(privKey, pubKey []byte) ([]byte, error) {
	priv := secp.PrivKeyFromBytes(privKey)
	pub, err := secp.ParsePubKey(pubKey)
	if err != nil {
		return nil, fmt.Errorf("parse pubkey: %w", err)
	}
	return secp.GenerateSharedSecret(priv, pub), nil
}

func deriveKeyMaterial(sharedSec []byte) []byte {
	salt := []byte("nip44-v2")
	info := []byte("nip44-key-material")
	hkdf := hkdf.New(sha256.New, sharedSec, salt, info)
	keyMaterial := make([]byte, keyMaterialSize)
	io.ReadFull(hkdf, keyMaterial)
	return keyMaterial
}

func Encrypt(plaintext string, privateKey, publicKey []byte) (string, error) {
	sharedSec, err := sharedSecretECDH(privateKey, publicKey)
	if err != nil {
		return "", fmt.Errorf("ecdh: %w", err)
	}

	keyMaterial := deriveKeyMaterial(sharedSec)
	encKey := keyMaterial[:keySize]
	nonce := keyMaterial[keySize : keySize+nonceSize]
	iv := keyMaterial[keySize+nonceSize:]

	aead, err := chacha20poly1305.New(encKey)
	if err != nil {
		return "", fmt.Errorf("new cipher: %w", err)
	}

	ciphertext := aead.Seal(nil, iv, []byte(plaintext), nil)

	mac := computeMAC(encKey, ciphertext, nonce, iv)

	payload := make([]byte, 1+nonceSize+len(ciphertext)+len(mac))
	payload[0] = nip44Version
	copy(payload[1:], nonce)
	copy(payload[1+nonceSize:], ciphertext)
	copy(payload[1+nonceSize+len(ciphertext):], mac)

	return base64.StdEncoding.EncodeToString(payload), nil
}

func Decrypt(ciphertextB64 string, privateKey, publicKey []byte) (string, error) {
	payload, err := base64.StdEncoding.DecodeString(ciphertextB64)
	if err != nil {
		return "", fmt.Errorf("base64 decode: %w", err)
	}
	if len(payload) < 1 {
		return "", fmt.Errorf("payload too short")
	}
	if payload[0] != nip44Version {
		return "", fmt.Errorf("unsupported version: %d", payload[0])
	}

	minLen := 1 + nonceSize + 1 + 16
	if len(payload) < minLen {
		return "", fmt.Errorf("payload too short for v2")
	}

	nonce := payload[1 : 1+nonceSize]
	ciphertext := payload[1+nonceSize : len(payload)-16]
	mac := payload[len(payload)-16:]

	sharedSec, err := sharedSecretECDH(privateKey, publicKey)
	if err != nil {
		return "", fmt.Errorf("ecdh: %w", err)
	}
	keyMaterial := deriveKeyMaterial(sharedSec)
	encKey := keyMaterial[:keySize]
	iv := keyMaterial[keySize+nonceSize:]

	expectedMAC := computeMAC(encKey, ciphertext, nonce, iv)
	if !hmac.Equal(mac, expectedMAC) {
		return "", fmt.Errorf("MAC mismatch: tampered ciphertext")
	}

	aead, err := chacha20poly1305.New(encKey)
	if err != nil {
		return "", fmt.Errorf("new cipher: %w", err)
	}

	plaintext, err := aead.Open(nil, iv, ciphertext, nil)
	if err != nil {
		return "", fmt.Errorf("decrypt: %w", err)
	}

	return string(plaintext), nil
}

func computeMAC(key, ciphertext, nonce, iv []byte) []byte {
	macInput := make([]byte, 0, len(nonce)+len(ciphertext)+len(iv))
	macInput = append(macInput, nonce...)
	macInput = append(macInput, ciphertext...)
	macInput = append(macInput, iv...)
	mac := hmac.New(sha256.New, key)
	mac.Write(macInput)
	return mac.Sum(nil)[:16]
}

func GenerateKey() (priv, pub []byte, err error) {
	privateKey, err := secp.GeneratePrivateKey()
	if err != nil {
		return nil, nil, err
	}
	return privateKey.Serialize(), privateKey.PubKey().SerializeCompressed(), nil
}

func RandomBytes(n int) ([]byte, error) {
	b := make([]byte, n)
	_, err := rand.Read(b)
	return b, err
}

File: web/src/crypto/nip44.ts
import { secp256k1 } from '@noble/curves/secp256k1'
import { sha256 } from '@noble/hashes/sha256'
import { hmac } from '@noble/hashes/hmac'
import { hkdf } from '@noble/hashes/hkdf'
import { chacha20poly1305 } from '@noble/ciphers/chacha'
import { bytesToHex, hexToBytes } from '@noble/hashes/utils'
import { toCompressedPubKey } from './keys'

const NIP44_VERSION = 0x02
const NONCE_SIZE = 32
const KEY_SIZE = 32
const IV_SIZE = 12
const KEY_MATERIAL_SIZE = KEY_SIZE + NONCE_SIZE + IV_SIZE

export function deriveSharedSecret(
  privateKey: Uint8Array,
  publicKey: Uint8Array,
): Uint8Array {
  const pk = publicKey.length === 32 ? toCompressedPubKey(publicKey) : publicKey
  const ss = secp256k1.getSharedSecret(privateKey, pk)
  return ss.subarray(1, 33)
}

function deriveKeyMaterial(sharedSecret: Uint8Array): Uint8Array {
  const salt = new TextEncoder().encode('nip44-v2')
  const info = new TextEncoder().encode('nip44-key-material')
  return hkdf(sha256, sharedSecret, salt, info, KEY_MATERIAL_SIZE)
}

export function encrypt(
  plaintext: string,
  privateKey: Uint8Array,
  publicKey: Uint8Array,
): string {
  const sharedSecret = deriveSharedSecret(privateKey, publicKey)
  const keyMaterial = deriveKeyMaterial(sharedSecret)

  const encKey = keyMaterial.subarray(0, KEY_SIZE)
  const nonce = keyMaterial.subarray(KEY_SIZE, KEY_SIZE + NONCE_SIZE)
  const iv = keyMaterial.subarray(KEY_SIZE + NONCE_SIZE)

  const cipher = chacha20poly1305(encKey, iv)
  const ciphertext = cipher.encrypt(new TextEncoder().encode(plaintext))

  const mac = computeMAC(encKey, ciphertext, nonce, iv)

  const payload = new Uint8Array(1 + nonce.length + ciphertext.length + mac.length)
  payload[0] = NIP44_VERSION
  payload.set(nonce, 1)
  payload.set(ciphertext, 1 + nonce.length)
  payload.set(mac, 1 + nonce.length + ciphertext.length)

  return bytesToHex(payload)
}

export function decrypt(
  ciphertextHex: string,
  privateKey: Uint8Array,
  publicKey: Uint8Array,
): string {
  const payload = hexToBytes(ciphertextHex)
  if (payload.length < 1) throw new Error('payload too short')
  if (payload[0] !== NIP44_VERSION) throw new Error(`unsupported version: ${payload[0]}`)

  const nonce = payload.subarray(1, 1 + NONCE_SIZE)
  const ciphertext = payload.subarray(1 + NONCE_SIZE, payload.length - 16)
  const mac = payload.subarray(payload.length - 16)

  const sharedSecret = deriveSharedSecret(privateKey, publicKey)
  const keyMaterial = deriveKeyMaterial(sharedSecret)

  const encKey = keyMaterial.subarray(0, KEY_SIZE)
  const iv = keyMaterial.subarray(KEY_SIZE + NONCE_SIZE)

  const expectedMAC = computeMAC(encKey, ciphertext, nonce, iv)
  if (!constantTimeEqual(mac, expectedMAC)) {
    throw new Error('MAC mismatch: tampered ciphertext')
  }

  const cipher = chacha20poly1305(encKey, iv)
  const plaintext = cipher.decrypt(ciphertext)
  return new TextDecoder().decode(plaintext)
}

function computeMAC(
  key: Uint8Array,
  ciphertext: Uint8Array,
  nonce: Uint8Array,
  iv: Uint8Array,
): Uint8Array {
  const input = new Uint8Array(nonce.length + ciphertext.length + iv.length)
  input.set(nonce, 0)
  input.set(ciphertext, nonce.length)
  input.set(iv, nonce.length + ciphertext.length)
  return hmac(sha256, key, input).subarray(0, 16)
}

function constantTimeEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false
  let diff = 0
  for (let i = 0; i < a.length; i++) {
    diff |= a[i] ^ b[i]
  }
  return diff === 0
}
